src/ui/suspense/styleCaches.ts

```diff
@@ -0,0 +1,135 @@
+  getValueIfCached: getAppliedRulesIfCached,
+} = createGenericCache<
+  [
+    client: ProtocolClient,
+    replayClient: ReplayClientInterface,
+    sessionId: string,
+    pauseId: PauseId,
+    nodeId: string
+  ],
+  WiredAppliedRule[]
+>(
+  async (client, replayClient, sessionId, pauseId, nodeId) => {
+    const { rules, data } = await client.CSS.getAppliedRules({ node: nodeId }, sessionId, pauseId);
+
+    const uniqueRules = uniqBy(rules, rule => `${rule.rule}|${rule.pseudoElement}`);
+
+    preCacheObjects(pauseId, data.objects ?? []);
+
+    const stylePromises: Promise<ProtocolObject>[] = [];
+
+    const rulePreviews = await Promise.all(
+      uniqueRules.map(async appliedRule => {
+        return getObjectWithPreviewHelper(replayClient, pauseId, appliedRule.rule);
+      })
+    );
+
+    for (let ruleObject of rulePreviews) {
+      if (ruleObject.preview?.rule?.style) {
+        stylePromises.push(
+          getObjectWithPreviewHelper(replayClient, pauseId, ruleObject.preview.rule.style)
+        );
+      }
+
+      if (ruleObject.preview?.rule?.parentStyleSheet) {
+        stylePromises.push(
+          getObjectWithPreviewHelper(
+            replayClient,
+            pauseId,
+            ruleObject.preview?.rule?.parentStyleSheet
+          )
+        );
+      }
+    }
+
+    if (stylePromises.length) {
+      await Promise.all(stylePromises);
+    }
+
+    const wiredRules: WiredAppliedRule[] = uniqueRules.map((appliedRule, i) => {
+      return {
+        rule: new RuleFront(pauseId, rulePreviews[i]),
+        pseudoElement: appliedRule.pseudoElement,
+      };
+    });
+    return wiredRules;
+  },
+  (client, replayClient, sessionId, pauseId, nodeId) => `${pauseId}|${nodeId}`
+);
+
+export const {
+  getValueSuspense: getComputedStyleSuspense,
+  getValueAsync: getComputedStyleAsync,
+  getValueIfCached: getComputedStyleIfCached,
+} = createGenericCache<
+  [client: ProtocolClient, sessionId: string, pauseId: PauseId, nodeId: string],
+  Map<string, string> | undefined
+>(
+  async (client, sessionId, pauseId, nodeId) => {
+    try {
+      const { computedStyle } = await client.CSS.getComputedStyle(
+        {
+          node: nodeId,
+        },
+        sessionId,
+        pauseId
+      );
+
+      const computedStyleMap = new Map();
+      for (const { name, value } of computedStyle) {
+        computedStyleMap.set(name, value);
+      }
+
+      return computedStyleMap;
+    } catch (err) {
+      return;
+    }
+  },
+  (client, sessionId, pauseId, nodeId) => `${pauseId}|${nodeId}`
+);
+
+export const {
+  getValueSuspense: getBoundingRectSuspense,
+  getValueAsync: getBoundingRectAsync,
+  getValueIfCached: getBoundingRectIfCached,
+} = createGenericCache<
+  [client: ProtocolClient, sessionId: string, pauseId: PauseId, nodeId: string],
+  DOMRect | undefined
+>(
+  async (client, sessionId, pauseId, nodeId) => {
+    try {
+      const { rect } = await client.DOM.getBoundingClientRect(
```

Comment:
Using the protocol client directly (rather than the Replay client) limits our ability to write new Playwright style tests for this stuff. That's probably _fine_ though, just maybe something it would be nice to eventually have the option to do :)

Reply:
Hmm. Last time I tried adding more methods to `ReplayClient`, you'd said you were trying to keep that minimal, so I was under the impression I should avoid doing that.  Also since the existing code just called `client.DOM.whatever()` directly, I was copying that logic.

Reply:
I think you're referring to different kinds of methods. `ReplayClient` exists to proxy calls to the Replay protocol (specifically because this allows us to mock out data for e2e tests).

It's not a great place to add utility functions.